### PR TITLE
protype of metrics and files to modify

### PR DESCRIPTION
**Pull request by edyoshikun**
This PR adds:
- [ ] dataloader for getting inference vs experimental fluoresnce metrics
- [ ] intensity metrcs
- [ ] biological metrics
- [ ] segmentation metrics
- [ ] logic to run metrics via VisCy CLI

### CODE REVIEW

File: viscy/data/segmentation.py [hunk 1]
```diff
@@ -55,26 +76,28 @@ def __len__(self) -> int:
...
-        target = torch.from_numpy(
-            target_img[t, self.target_channel, self.target_z_slice].astype(np.int16)
-        )
+        _pred = pred_img[t, self.pred_channel, self.pred_z_slice]
+        _target = target_img[t, self.target_channel, self.target_z_slice]
+        if self.dtype is not None:
+            _pred = _pred.astype(self.dtype)
+            _target = _target.astype(self.dtype)
+        pred = torch.from_numpy(_pred.astype(self.dtype))
+        target = torch.from_numpy(_target.astype(self.dtype))
```
Comment (ziw-liu) [inline viscy/data/segmentation.py:85]: Why do this twice?

File: applications/DynaCell/demo_script.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+"""
+This script is a demo script for the DynaCell application.
+It loads the ome-zarr 0.4v format, calculates metrics and saves the results as csv files
```
Comment (ziw-liu) [inline applications/DynaCell/demo_script.py:3]:
```suggestion
It loads the ome-zarr v0.4 format, calculates metrics and saves the results as csv files
```

File: applications/DynaCell/demo_script.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+from viscy.data.dynacell import DynaCellDataBase, DynaCellDataModule
+from viscy.trainer import Trainer
+from viscy.translation.evaluation import IntensityMetrics, SegmentationMetrics
+
+# Set float32 matmul precision for better performance on Tensor Cores
+torch.set_float32_matmul_precision("high")
+
+csv_database_path = Path(
+    "/home/eduardo.hirata/repos/viscy/applications/DynaCell/dynacell_summary_table.csv"
+).expanduser()
```
Comment (ziw-liu) [inline applications/DynaCell/demo_script.py:24]: `Path.home() / "rel/path"` is likely what you want.

File: viscy/data/dynacell.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            {
+                "cell_type": self.cell_type,
+                "organelle": self.organelle,
+                "infection_condition": self.infection_condition,
+            }
+        )
+        return sample
+
+
+class DynaCellDataBase:
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:64]: Capital `XxxBase` reads like the name of a base class. Also this is not a databse with a runtime, but a dataframe normalizer.

File: viscy/data/dynacell.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        # Select the portion of the database that matches the provided criteria
+        self._filtered_db = self.database[
+            self.database["Cell type"].isin(self.cell_types)
+            & self.database["Organelle"].isin(self.organelles)
+            & self.database["Infection"].isin(self.infection_conditions)
+        ].copy()
+
+        # Extract zarr store paths
+        self._filtered_db["Zarr path"] = self._filtered_db["Path"].apply(
+            lambda x: Path(*Path(x).parts[:-3])
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:95]:
```suggestion
            lambda x: x.parent.parent.parent
```

File: viscy/data/dynacell.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        self.cell_types_per_store = self._filtered_db["Cell type"].values.tolist()
+        self.organelles_per_store = self._filtered_db["Organelle"].values.tolist()
+        self.infection_per_store = self._filtered_db["Infection"].values.tolist()
+
+    def __getitem__(self, idx) -> dict:
+        return {
+            "zarr_path": self.zarr_paths[idx],
+            "cell_type": self.cell_types_per_store[idx],
+            "organelle": self.organelles_per_store[idx],
+            "infection_condition": self.infection_per_store[idx],
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:110]: Why are these first converted to lists?

File: viscy/data/dynacell.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            )
+
+        # Create datasets
+        datasets = []
+        for i in range(len(self.target_database)):
+            target_data = self.target_database[i]
+            pred_data = self.pred_database[i]
+
+            # Ensure target and prediction metadata match
+            self._validate_matching_metadata(target_data, pred_data, i)
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:151]: How long is this loop?

File: viscy/data/dynacell.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+                f"Organelle mismatch at index {idx}: "
+                f"target={target_data['organelle']}, pred={pred_data['organelle']}"
+            )
+
+        # Check infection condition
+        if target_data["infection_condition"] != pred_data["infection_condition"]:
+            raise ValueError(
+                f"Infection condition mismatch at index {idx}: "
+                f"target={target_data['infection_condition']}, pred={pred_data['infection_condition']}"
+            )
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:192]: This can be a loop over the string keys.

File: viscy/data/dynacell.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        return DataLoader(
+            self.test_dataset,
+            batch_size=self.batch_size,
+            num_workers=self.num_workers,
+            collate_fn=self._custom_collate,
+        )
+
+    def _custom_collate(self, batch):
+        """Custom collate function that preserves metadata strings."""
+        # Extract metadata from first element in batch
```
Comment (ziw-liu) [inline viscy/data/dynacell.py:211]: Is this assuming batch size of 1?

File: viscy/data/segmentation.py [hunk 10]
```diff
@@ -14,25 +14,46 @@
 _logger = logging.getLogger("lightning.pytorch")
 
 
-class SegmentationDataset(Dataset):
+class TargetPredictionDataset(Dataset):
+    """
+    A PyTorch Dataset providing paired target and prediction images / volumes from OME-Zarr
+    datasets.
+
+    Attributes:
```
Comment (ziw-liu) [inline viscy/data/segmentation.py:22]: Please use numpy-style docstring: #133

File: viscy/data/typing.py [hunk 11]
```diff
@@ -64,6 +64,16 @@ class SegmentationSample(TypedDict):
...
 
 
+class DynaCellSample(SegmentationSample):
+    """
+    DynaCell sample type for mini-batches.
+    """
+
+    cell_type: str
+    organelle: str
+    infection: str
```
Comment (ziw-liu) [inline viscy/data/typing.py:74]: This key is not the same as the CSV column name?

File: viscy/trainer.py [hunk 12]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+            Cell types to include for prediction dataset, by default None (all available)
+        pred_organelles : list[str], optional
+            Organelles to include for prediction dataset, by default None (all available)
+        pred_infection_conditions : list[str], optional
+            Infection conditions to include for prediction dataset, by default None (all available)
+        batch_size : int, optional
+            Batch size for processing, by default 1
+        num_workers : int, optional
+            Number of workers for data loading, by default 0
+        version : str, optional
```
Comment (ziw-liu) [inline viscy/trainer.py:195]: What version is this?

File: viscy/trainer.py [hunk 13]
```diff
@@ -128,3 +133,240 @@ def precompute(
             include_wells=include_wells,
             exclude_fovs=exclude_fovs,
         )
+
+    def compute_dynacell_metrics(
+        self,
+        target_database: Path,
+        pred_database: Path,
+        output_dir: Path,
+        method: str = "intensity",
```
Comment (ziw-liu) [inline viscy/trainer.py:142]: Can this be all supported literals rather than a free string? Then lightning CLI can check the input type.

File: viscy/trainer.py [hunk 14]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+
+        # Unique name based on method and timestamp
+        run_name = f"{method}_{timestamp}"
+
+        # Create logger
+        _logger.info(f"Creating logger for run '{run_name}' with version '{version}'")
+        logger = CSVLogger(save_dir=method_dir, name=run_name, version=version)
+
+        # Create trainer
+        trainer = Trainer(logger=logger)
```
Comment (ziw-liu) [inline viscy/trainer.py:342]: `self` should already be a child of `Trainer`.

File: viscy/trainer.py [hunk 15]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+
+        # Select and run appropriate metrics
+        if method == "segmentation2D":
```
Comment (ziw-liu) [inline viscy/trainer.py:345]: I'd appreciate CamelCase or snake_case names.

File: viscy/trainer.py [hunk 16]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+            _logger.info("Running segmentation metrics...")
+            metrics_module = SegmentationMetrics()
+        else:  # intensity
```
Comment (ziw-liu) [inline viscy/trainer.py:348]: Should be an `elif` since the input is not type-checked.

File: viscy/trainer.py [hunk 17]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+            metrics_df = pd.read_csv(metrics_file)
+            _logger.info(f"Metrics saved to: {metrics_file}")
+            _logger.info(f"Computed {len(metrics_df)} metric rows")
+
+            # Display columns in the metrics
+            _logger.info(f"Metrics columns: {metrics_df.columns.tolist()}")
+
+            # Display a preview of the metrics
+            if not metrics_df.empty:
+                _logger.info(f"Metrics preview:\n{metrics_df.head().to_string()}")
```
Comment (ziw-liu) [inline viscy/trainer.py:368]: Are these redundant? Also maybe try repr to get pretty-printing.

File: viscy/trainer.py [hunk 18]
```diff
@@ -128,3 +133,240 @@ def precompute(
...
+        else:
+            _logger.warning(f"No metrics file found at {metrics_file}")
+
+        return metrics_file if metrics_file.exists() else None
```
Comment (ziw-liu) [inline viscy/trainer.py:372]: Maybe return None early in the previous branch than checking it again.

File: viscy/translation/evaluation.py [hunk 19]
```diff
@@ -12,43 +15,254 @@
...
-class SegmentationMetrics2D(LightningModule):
-    """Test runner for 2D segmentation."""
+class SegmentationMetrics(LightningModule):
+    """Test runner for segmentation that handles both 2D and 3D data.
 
-    def __init__(self, aggregate_epoch: bool = False) -> None:
+    Parameters
+    ----------
+    mode : str, optional
+        Mode of operation, can be "auto", "2D", or "3D", by default "auto".
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:24]: Can these be in the signature?

File: viscy/translation/evaluation.py [hunk 20]
```diff
@@ -12,43 +15,254 @@
...
+        # Common preprocessing for both modes
         pred_binary = pred > 0
         target_binary = target > 0
+
+        if current_mode == "2D":
+            self._compute_2d_metrics(pred, target, pred_binary, target_binary, batch)
+        else:  # 3D mode
+            self._compute_3d_metrics(pred, target, pred_binary, target_binary, batch)
+
+    def _compute_2d_metrics(self, pred, target, pred_binary, target_binary, batch):
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:85]: Please add type hints.

File: viscy/translation/evaluation.py [hunk 21]
```diff
@@ -12,43 +15,254 @@
...
+            "mAR_100": coco_metrics["mar_100"],
+        }
+
+        # Add metadata if available
+        if "cell_type" in batch:
+            metrics_dict["cell_type"] = batch["cell_type"]
+        if "organelle" in batch:
+            metrics_dict["organelle"] = batch["organelle"]
+        if "infection_condition" in batch:
+            metrics_dict["infection_condition"] = batch["infection_condition"]
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:116]: These should be isolated in a dynacell-specific method.

File: viscy/translation/evaluation.py (hunk 20 above)
Comment (ziw-liu) [reply viscy/translation/evaluation.py:85]: Also `batch` contains `pred` and `target`...

File: viscy/translation/evaluation.py (hunk 21 above)
Comment (ziw-liu) [reply viscy/translation/evaluation.py:116]: These are also repeated below so the said method should be called in `test_step`.

File: viscy/translation/evaluation.py [hunk 22]
```diff
@@ -12,43 +15,254 @@
...
+        ## Measuring Panoptic Quality
+        evaluator = Panoptica_Evaluator(
+            expected_input=InputType.UnmatchedInstancePair,
+            instance_metrics=[Metric.DSC, Metric.IoU],
+            decision_metric=Metric.DSC,
+            decision_threshold=0.5,
+            log_times=True,
+        )
+        result = evaluator.evaluate(pred, target, verbose=False)
+        result = result.to_dict()
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:138]: This wrapper should be moved to its own function, similar to `mean_average_precision` and have similar tests to ensure correctness.

File: viscy/translation/evaluation.py [hunk 23]
```diff
@@ -12,43 +15,254 @@
...
+            **result,
+        }
+
+        # Add metadata if available
+        if "cell_type" in batch:
+            metrics_dict["cell_type"] = batch["cell_type"]
+        if "organelle" in batch:
+            metrics_dict["organelle"] = batch["organelle"]
+        if "infection_condition" in batch:
+            metrics_dict["infection_condition"] = batch["infection_condition"]
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:156]: https://github.com/mehta-lab/VisCy/pull/242/files#r2103601552

File: viscy/translation/evaluation.py [hunk 24]
```diff
@@ -12,43 +15,254 @@
...
+
+        self.logger.log_metrics(metrics_dict)
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:158]: This is repeated in 2D and 3D. Should return the metrics (as the 'compute' in the method name would suggests) and let the calling function log once.

File: viscy/translation/evaluation.py [hunk 25]
```diff
@@ -12,43 +15,254 @@
...
+        aggregate_epoch: bool = False,
+    ) -> None:
+        super().__init__()
+        self.metrics = metrics
+        self.aggregate_epoch = aggregate_epoch
+        self._validate_metrics()
+
+    def _validate_metrics(self):
+        """Validate the metrics parameter."""
+        valid_metrics = ["mae", "mse", "ssim", "ms_ssim", "pearson", "cosine"]
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:196]: These should be documented in the signature of `__init__`, i.e. `metrics: list[Literal["mae", "mse", "ssim", "ms_ssim", "pearson", "cosine"]]`.

File: viscy/translation/evaluation.py [hunk 26]
```diff
@@ -12,43 +15,254 @@
...
+        for metric in self.metrics:
+            if metric not in valid_metrics:
+                raise ValueError(f"Metric '{metric}' not in {valid_metrics}")
+
+    def test_step(self, batch, batch_idx: int) -> None:
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:201]: What is the type of `batch`?

File: viscy/translation/evaluation.py [hunk 27]
```diff
@@ -12,43 +15,254 @@
...
+        """Compute intensity metrics between prediction and target."""
+        from torchmetrics.functional import (
+            cosine_similarity,
+            mean_absolute_error,
+            mean_squared_error,
+            pearson_corrcoef,
+            structural_similarity_index_measure,
+        )
+
+        from viscy.translation.evaluation_metrics import ms_ssim_25d
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:211]: Why import separately from other metrics?

File: viscy/translation/evaluation.py [hunk 28]
```diff
@@ -12,43 +15,254 @@
...
+            "time": batch["time_idx"][0] if "time_idx" in batch else -1,
+        }
+
+        # Add metadata if available
+        if "cell_type" in batch:
+            metrics_dict["cell_type"] = batch["cell_type"]
+        if "organelle" in batch:
+            metrics_dict["organelle"] = batch["organelle"]
+        if "infection_condition" in batch:
+            metrics_dict["infection_condition"] = batch["infection_condition"]
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:228]: https://github.com/mehta-lab/VisCy/pull/242/files#r2103601552

File: viscy/translation/evaluation.py [hunk 29]
```diff
@@ -12,43 +15,254 @@
...
+                            else target[:, :, target.shape[2] // 2]
+                        ),
+                    )
+                else:  # 2D case
+                    metrics_dict["ssim"] = structural_similarity_index_measure(
+                        pred, target
+                    )
+            elif metric == "ms_ssim":
+                if pred.ndim > 1:
+                    metrics_dict["ms_ssim"] = ms_ssim_25d(pred, target)
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:257]: This function is defined only for use in a loss function. (I made it up...)

File: viscy/translation/evaluation.py [hunk 30]
```diff
@@ -12,43 +15,254 @@
...
+                    metrics_dict["ssim"] = structural_similarity_index_measure(
+                        (
+                            pred.squeeze(2)
+                            if pred.shape[2] == 1
+                            else pred[:, :, pred.shape[2] // 2]
+                        ),
+                        (
+                            target.squeeze(2)
+                            if target.shape[2] == 1
+                            else target[:, :, target.shape[2] // 2]
```
Comment (ziw-liu) [inline viscy/translation/evaluation.py:248]: Potentially dangerous: this could compute SSIM on a different region than the other metrics.

File: viscy/data/dynacell.py (hunk 7 above)
Comment (edyoshikun) [reply viscy/data/dynacell.py:166]: for now we don't parallelize. Are you thinking on spitting it as batches?

File: viscy/data/dynacell.py (hunk 9 above)
Comment (edyoshikun) [reply viscy/data/dynacell.py:227]: yes...

File: viscy/trainer.py (hunk 13 above)
Comment (edyoshikun) [reply viscy/trainer.py:142]: yes, we will update this once we have all the possible ones. We might make it it's own list for the organelles. thanks

File: viscy/data/dynacell.py (hunk 9 above)
Comment (ziw-liu) [reply viscy/data/dynacell.py:227]: Then the constructor should error if it isn't 1.

**Comment by edyoshikun** [comment]
drifted and stale
